NLL / MIR-borrowck regression in liballoc

When compiling rustc with NLL (full command: `RUSTFLAGS_STAGE_NOT_0='-Z borrowck=mir -Z nll -Z two-phase-borrows' ./x.py build` ), `liballoc` fails to build with the following errors:

```
error[E0311]: the parameter type `K` may not live long enough
   --> liballoc/btree/map.rs:158:19
    |
158 |             match node.force() {
    |                   ^^^^
    |
    = help: consider adding an explicit lifetime bound for `K`
note: the parameter type `K` must be valid for the anonymous lifetime #1 defined on the function body at 152:9...
   --> liballoc/btree/map.rs:152:9
    |
152 | /         fn clone_subtree<K: Clone, V: Clone>(node: node::NodeRef<marker::Immut,
153 | |                                                                  K,
154 | |                                                                  V,
155 | |                                                                  marker::LeafOrInternal>)
...   |
213 | |             }
214 | |         }
    | |_________^

error[E0311]: the parameter type `V` may not live long enough
   --> liballoc/btree/map.rs:158:19
    |
158 |             match node.force() {
    |                   ^^^^
    |
    = help: consider adding an explicit lifetime bound for `V`
note: the parameter type `V` must be valid for the anonymous lifetime #1 defined on the function body at 152:9...
   --> liballoc/btree/map.rs:152:9
    |
152 | /         fn clone_subtree<K: Clone, V: Clone>(node: node::NodeRef<marker::Immut,
153 | |                                                                  K,
154 | |                                                                  V,
155 | |                                                                  marker::LeafOrInternal>)
...   |
213 | |             }
214 | |         }
    | |_________^

error[E0311]: the parameter type `K` may not live long enough
    --> liballoc/btree/map.rs:1046:37
     |
1046 |             if let Internal(node) = node.force() {
     |                                     ^^^^
     |
     = help: consider adding an explicit lifetime bound for `K`
note: the parameter type `K` must be valid for the anonymous lifetime #1 defined on the function body at 1043:9...
    --> liballoc/btree/map.rs:1043:9
     |
1043 | /         fn dfs<K, V>(node: NodeRef<marker::Immut, K, V, marker::LeafOrInternal>) -> usize {
1044 | |             let mut res = node.len();
1045 | |
1046 | |             if let Internal(node) = node.force() {
...    |
1061 | |             res
1062 | |         }
     | |_________^

error[E0311]: the parameter type `V` may not live long enough
    --> liballoc/btree/map.rs:1046:37
     |
1046 |             if let Internal(node) = node.force() {
     |                                     ^^^^
     |
     = help: consider adding an explicit lifetime bound for `V`
note: the parameter type `V` must be valid for the anonymous lifetime #1 defined on the function body at 1043:9...
    --> liballoc/btree/map.rs:1043:9
     |
1043 | /         fn dfs<K, V>(node: NodeRef<marker::Immut, K, V, marker::LeafOrInternal>) -> usize {
1044 | |             let mut res = node.len();
1045 | |
1046 | |             if let Internal(node) = node.force() {
...    |
1061 | |             res
1062 | |         }
     | |_________^
```

--- a/src/liballoc/collections/btree/map.rs
+++ b/src/liballoc/collections/btree/map.rs
@@ -149,12 +149,11 @@ unsafe impl<#[may_dangle] K, #[may_dangle] V> Drop for BTreeMap<K, V> {
 #[stable(feature = "rust1", since = "1.0.0")]
 impl<K: Clone, V: Clone> Clone for BTreeMap<K, V> {
     fn clone(&self) -> BTreeMap<K, V> {
-        fn clone_subtree<K: Clone, V: Clone>(node: node::NodeRef<marker::Immut,
-                                                                 K,
-                                                                 V,
-                                                                 marker::LeafOrInternal>)
-                                             -> BTreeMap<K, V> {
-
+        fn clone_subtree<'a, K: Clone, V: Clone>(
+            node: node::NodeRef<marker::Immut<'a>, K, V, marker::LeafOrInternal>
+        ) -> BTreeMap<K, V>
+        where K: 'a, V: 'a,
+        {
             match node.force() {
                 Leaf(leaf) => {
                     let mut out_tree = BTreeMap {
@@ -1080,7 +1079,11 @@ impl<K: Ord, V> BTreeMap<K, V> {
 
     /// Calculates the number of elements if it is incorrect.
     fn recalc_length(&mut self) {
-        fn dfs<K, V>(node: NodeRef<marker::Immut, K, V, marker::LeafOrInternal>) -> usize {
+        fn dfs<'a, K, V>(
+            node: NodeRef<marker::Immut<'a>, K, V, marker::LeafOrInternal>
+        ) -> usize
+        where K: 'a, V: 'a
+        {
             let mut res = node.len();
 
             if let Internal(node) = node.force() {

